server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java

```diff
@@ -0,0 +1,578 @@
+                    continue;
+                }
+
+                if (eligibleNodes.isEmpty()) {
+                    break;
+                }
+
+                Set<String> nodesCheckedForShard = new HashSet<>();
+                while (!eligibleNodes.isEmpty()) {
+                    RoutingNode targetNode = eligibleNodes.poll();
+                    Decision currentShardDecision = allocation.deciders().canAllocate(shard, targetNode, allocation);
+
+                    if (currentShardDecision.type() == Decision.Type.YES) {
+                        if (logger.isDebugEnabled()) {
+                            logger.debug(
+                                "Moving shard: {} from node: [{}] to node: [{}]",
+                                shardShortSummary(shard),
+                                shard.currentNodeId(),
+                                targetNode.nodeId()
+                            );
+                        }
+                        routingNodes.relocateShard(
+                            shard,
+                            targetNode.nodeId(),
+                            allocation.clusterInfo().getShardSize(shard, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE),
+                            allocation.changes()
+                        );
+                        eligibleNodes.offer(targetNode);
+                        break;
+                    } else {
+                        if (logger.isTraceEnabled()) {
+                            logger.trace(
+                                "Cannot move shard: {} to node: [{}]. Decisions: [{}]",
+                                shardShortSummary(shard),
+                                targetNode.nodeId(),
+                                currentShardDecision.getDecisions()
+                            );
+                        }
+
+                        Decision nodeLevelDecision = allocation.deciders().canAllocateAnyShardToNode(targetNode, allocation);
+                        if (nodeLevelDecision.type() == Decision.Type.YES) {
+                            logger.debug("Node: [{}] can still accept shards. Adding it back to the queue.", targetNode.nodeId());
+                            eligibleNodes.offer(targetNode);
+                            nodesCheckedForShard.add(targetNode.nodeId());
+                        } else {
+                            logger.debug("Node: [{}] cannot accept any more shards. Removing it from queue.", targetNode.nodeId());
+                        }
+
+                        // Break out if all nodes in the queue have been checked for this shard
+                        if (eligibleNodes.stream().allMatch(rn -> nodesCheckedForShard.contains(rn.nodeId()))) {
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Performs heuristic based balancing for remote shards within the cluster.
```

Comment:
Can you expand a bit here? For example, what is the heuristic being used here? Doesn't need to be super detailed, just a high level description of exactly what is being done.

Reply:
Done!